src/utils/messaging-outbox/src/services/outbox_transactional_processor.rs

```diff
@@ -320,35 +321,80 @@ impl ProducerRelayJob {
+                    if failing_consumer_names.contains(consumer_name) {
+                        continue;
+                    }
+
+                    // Skip consumers, which are already beyond this message ID
                     let boundary_id = consumption_boundaries
                         .get(consumer_name)
                         .copied()
                         .unwrap_or_else(|| OutboxMessageID::new(0));
                     if boundary_id < message.message_id {
+                        // Non-failing consumer which hasn't seen this message is a task to execute
                         consumer_tasks.push((consumer_name.as_str(), &message));
                     }
                 }
 
-                // Consume concurrently
-                // TODO: error recovery
-                use futures::{StreamExt, TryStreamExt};
-                futures::stream::iter(consumer_tasks)
-                    .map(Ok)
-                    .try_for_each_concurrent(
-                        /* limit */ None,
-                        |(consumer_name, message)| async move {
-                            self.invoke_consumer(consumer_name, message).await?;
-                            Ok(())
-                        },
-                    )
-                    .await?;
+                #[derive(Error, Debug)]
+                #[error("Outbox message consumption error")]
+                struct OutboxMessageConsumptionError {
+                    #[source]
+                    source: InternalError,
+                    consumer_name: String,
+                    message_id: OutboxMessageID,
+                }
+
+                // Consume message concurrently
+                let consumption_futures: Vec<_> = consumer_tasks
+                    .into_iter()
+                    .map(|(consumer_name, message)| async move {
+                        self.invoke_consumer(consumer_name, message)
+                            .await
+                            .map_err(|e| OutboxMessageConsumptionError {
+                                consumer_name: consumer_name.to_owned(),
+                                message_id: message.message_id,
+                                source: e,
+                            })
+                    })
+                    .collect();
+
+                // Report errors and disable failing consumers temporarily, as they cannot
+                // advance to the next messages without success of processing the current one
+                futures::future::join_all(consumption_futures)
+                    .await
+                    .into_iter()
+                    .filter(Result::is_err)
+                    .map(|e| e.err().unwrap())
+                    .for_each(|e: OutboxMessageConsumptionError| {
+                        tracing::error!(
+                            error=?e, message=?message,
+                            "Consuming outbox message FAILED"
```

Comment:
Same about format.

Maybe add `consumer_name` to the fields so it's easier to see without reading error debug output?

Also, it would be useful to explain what failing to consume a message means, e.g.:
```
Consuming outbox message failed - pausing further message processing for consumer until restart.
```

Reply:
`consumer_name` added

Reply:
Message improved

Reply:
Implemented pausing until restart

Reply:
https://github.com/kamu-data/kamu-node/issues/130

Ticket for metric.